File: go/11_tool_use.go
package main

/*
Example 10: Tool Use (Function Calling) - AI Agents with Groq + Tavily

Demonstrates:
- Groq function calling with llama-4-scout
- Tavily Search and Extract as tools
- Multi-step agent workflow

Note: Uses Go standard library only
For production, consider using official SDKs
*/

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
)

// Tool function definitions
func tavilySearch(query string) (map[string]interface{}, error) {
	fmt.Printf("[Tool] Executing tavily_search with query: %s\n", query)

	requestBody, _ := json.Marshal(map[string]interface{}{
		"query":          query,
		"max_results":    5,
		"include_answer": true,
	})

	req, err := http.NewRequest("POST", "https://api.tavily.com/search", bytes.NewBuffer(requestBody))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", os.Getenv("TAVILY_API_KEY")))

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	var result map[string]interface{}
	json.Unmarshal(body, &result)
	return result, nil
}

func tavilyExtract(url string) (map[string]interface{}, error) {
	fmt.Printf("[Tool] Executing tavily_extract for URL: %s\n", url)

	requestBody, _ := json.Marshal(map[string]interface{}{
		"urls":          []string{url},
		"extract_depth": "basic",
	})

	req, err := http.NewRequest("POST", "https://api.tavily.com/extract", bytes.NewBuffer(requestBody))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", os.Getenv("TAVILY_API_KEY")))

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	var result map[string]interface{}
	json.Unmarshal(body, &result)
	return result, nil
}

func executeTool(functionName string, arguments map[string]interface{}) (map[string]interface{}, error) {
	switch functionName {
	case "tavily_search":
		return tavilySearch(arguments["query"].(string))
	case "tavily_extract":
		return tavilyExtract(arguments["url"].(string))
	default:
		return nil, fmt.Errorf("unknown function: %s", functionName)
	}
}

func callGroq(messages []interface{}, tools []interface{}) (map[string]interface{}, error) {
	requestData := map[string]interface{}{
		"model":    "meta-llama/llama-4-scout-17b-16e-instruct",
		"messages": messages,
	}

	if tools != nil {
		requestData["tools"] = tools
		requestData["tool_choice"] = "auto"
	}

	requestBody, _ := json.Marshal(requestData)

	req, err := http.NewRequest("POST", "https://api.groq.com/openai/v1/chat/completions", bytes.NewBuffer(requestBody))
	if err != nil {
		return nil, err
	}

	req.Header.Set("Content-Type", "application/json")
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", os.Getenv("GROQ_API_KEY")))

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, _ := io.ReadAll(resp.Body)

	var result map[string]interface{}
	json.Unmarshal(body, &result)
	return result, nil
}

func main() {
	// Check for API keys
	if os.Getenv("GROQ_API_KEY") == "" || os.Getenv("TAVILY_API_KEY") == "" {
		fmt.Fprintln(os.Stderr, "Error: Both GROQ_API_KEY and TAVILY_API_KEY must be set")
		fmt.Fprintln(os.Stderr, "Get Groq key: https://console.groq.com")
		fmt.Fprintln(os.Stderr, "Get Tavily key: https://tavily.com")
		os.Exit(1)
	}

	fmt.Println("========================================")
	fmt.Println("Groq Tool Use - AI Agent Demo")
	fmt.Println("========================================\n")

	userQuery := "Search for the latest developments in quantum computing and provide a summary"

	fmt.Printf("User Query: %s\n\n", userQuery)
	fmt.Println("Step 1: Sending query to AI with tool definitions...\n")

	// Tool definitions
	tools := []interface{}{
		map[string]interface{}{
			"type": "function",
			"function": map[string]interface{}{
				"name":        "tavily_search",
				"description": "Search the web for current information using Tavily API",
				"parameters": map[string]interface{}{
					"type": "object",
					"properties": map[string]interface{}{
						"query": map[string]interface{}{
							"type":        "string",
							"description": "The search query",
						},
					},
					"required": []string{"query"},
				},
			},
		},
		map[string]interface{}{
			"type": "function",
			"function": map[string]interface{}{
				"name":        "tavily_extract",
				"description": "Extract content from a specific URL",
				"parameters": map[string]interface{}{
					"type": "object",
					"properties": map[string]interface{}{
						"url": map[string]interface{}{
							"type":        "string",
							"description": "The URL to extract content from",
						},
					},
					"required": []string{"url"},
				},
			},
		},
	}

	// Step 1: Initial request
	initialResponse, err := callGroq([]interface{}{
		map[string]interface{}{"role": "user", "content": userQuery},
	}, tools)

	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	// Check for tool calls
	choices := initialResponse["choices"].([]interface{})
	message := choices[0].(map[string]interface{})["message"].(map[string]interface{})
	toolCalls, hasToolCalls := message["tool_calls"]

	if !hasToolCalls || toolCalls == nil {
		fmt.Println("AI answered without using tools:\n")
		fmt.Println(message["content"])
		return
	}

	// Step 2: Execute tool calls
	fmt.Println("========================================")
	fmt.Println("AI decided to use tools!")
	fmt.Println("========================================\n")

	toolCallsList := toolCalls.([]interface{})
	fmt.Printf("Number of tool calls: %d\n\n", len(toolCallsList))

	var toolMessages []interface{}

	for _, toolCall := range toolCallsList {
		tc := toolCall.(map[string]interface{})
		function := tc["function"].(map[string]interface{})
		functionName := function["name"].(string)
		argumentsStr := function["arguments"].(string)

		fmt.Printf("Function: %s\n", functionName)
		fmt.Printf("Arguments: %s\n\n", argumentsStr)

		var arguments map[string]interface{}
		json.Unmarshal([]byte(argumentsStr), &arguments)

		result, err := executeTool(functionName, arguments)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error executing tool: %v\n", err)
			continue
		}

		resultJSON, _ := json.Marshal(result)

		toolMessages = append(toolMessages, map[string]interface{}{
			"role":         "tool",
			"tool_call_id": tc["id"],
			"content":      string(resultJSON),
		})

		fmt.Println("✓ Tool executed successfully\n")
	}

	// Step 3: Send results back to AI
	fmt.Println("Step 2: Sending tool results back to AI for final answer...\n")

	messages := []interface{}{
		map[string]interface{}{"role": "user", "content": userQuery},
		message,
	}
	messages = append(messages, toolMessages...)

	finalResponse, err := callGroq(messages, nil)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	// Display final response
	fmt.Println("========================================")
	fmt.Println("Final AI Response")
	fmt.Println("========================================\n")

	finalChoices := finalResponse["choices"].([]interface{})
	finalMessage := finalChoices[0].(map[string]interface{})["message"].(map[string]interface{})
	fmt.Println(finalMessage["content"])
	fmt.Println()

	// Display token usage
	fmt.Println("========================================")
	fmt.Println("Token Usage")
	fmt.Println("========================================\n")

	usage := finalResponse["usage"].(map[string]interface{})
	promptTokens := usage["prompt_tokens"].(float64)
	completionTokens := usage["completion_tokens"].(float64)
	totalTokens := usage["total_tokens"].(float64)

	fmt.Printf("Prompt tokens: %.0f\n", promptTokens)
	fmt.Printf("Completion tokens: %.0f\n", completionTokens)
	fmt.Printf("Total tokens: %.0f\n", totalTokens)

	cost := (promptTokens*0.11 + completionTokens*0.34) / 1000000
	fmt.Printf("Cost: $%.6f\n\n", cost)

	fmt.Println("========================================")
	fmt.Println("Summary")
	fmt.Println("========================================\n")
	fmt.Println("✓ AI agent successfully used tools")
	fmt.Printf("✓ Executed %d tool call(s)\n", len(toolCallsList))
	fmt.Println("✓ Retrieved real-time information")
	fmt.Println("✓ Generated informed response\n")
}

File: cpp/11_tool_use.cpp
/**
 * ============================================================================
 * Example 11: Tool Use - AI Agents with Function Calling
 * ============================================================================
 *
 * WHAT THIS DEMONSTRATES:
 *   - Giving AI the ability to call functions/tools
 *   - Building an AI agent that can take actions
 *   - Tool definition and execution loop
 *
 * WHAT YOU'LL LEARN:
 *   - OpenAI-compatible tool calling format
 *   - Agent loop pattern
 *   - Combining AI decisions with external actions
 *
 * COMPILE:
 *   g++ -std=c++17 -o 11_tool_use 11_tool_use.cpp -lcurl -I lib/
 *
 * RUN:
 *   ./11_tool_use
 *
 * ============================================================================
 */

#include <iostream>
#include <string>
#include <cstdlib>
#include <ctime>
#include <functional>
#include <map>
#include <curl/curl.h>
#include "json.hpp"

using json = nlohmann::json;

static size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    std::string *response = static_cast<std::string*>(userp);
    response->append(static_cast<char*>(contents), size * nmemb);
    return size * nmemb;
}

/* ============================================================================
 * TOOL IMPLEMENTATIONS
 * ============================================================================
 * These are the actual functions the AI can call.
 */

std::string get_current_time() {
    time_t now = time(nullptr);
    char buffer[64];
    strftime(buffer, sizeof(buffer), "%Y-%m-%d %H:%M:%S", localtime(&now));
    return buffer;
}

std::string calculate(double a, double b, const std::string &operation) {
    double result = 0;
    if (operation == "add") result = a + b;
    else if (operation == "subtract") result = a - b;
    else if (operation == "multiply") result = a * b;
    else if (operation == "divide") result = (b != 0) ? a / b : 0;

    return std::to_string(result);
}

std::string get_weather(const std::string &location) {
    /* In real app, this would call a weather API */
    return "Weather in " + location + ": Sunny, 22C (72F), humidity 45%";
}

int main() {
    /* Step 1: Get API key */
    const char *api_key_cstr = std::getenv("DEMETERICS_API_KEY");
    if (api_key_cstr == nullptr) {
        std::cerr << "Error: DEMETERICS_API_KEY not set\n";
        return 1;
    }
    std::string api_key(api_key_cstr);

    /* Step 2: Initialize */
    curl_global_init(CURL_GLOBAL_ALL);

    std::cout << "=== AI AGENT WITH TOOLS ===\n";
    std::cout << "The AI can call these tools:\n";
    std::cout << "  - get_current_time: Get current date/time\n";
    std::cout << "  - calculate: Perform math operations\n";
    std::cout << "  - get_weather: Get weather for a location\n\n";

    /* Step 3: Define the tools (functions) the AI can use */
    json tools = json::array({
        {
            {"type", "function"},
            {"function", {
                {"name", "get_current_time"},
                {"description", "Get the current date and time"},
                {"parameters", {
                    {"type", "object"},
                    {"properties", json::object()},
                    {"required", json::array()}
                }}
            }}
        },
        {
            {"type", "function"},
            {"function", {
                {"name", "calculate"},
                {"description", "Perform a mathematical calculation"},
                {"parameters", {
                    {"type", "object"},
                    {"properties", {
                        {"a", {{"type", "number"}, {"description", "First number"}}},
                        {"b", {{"type", "number"}, {"description", "Second number"}}},
                        {"operation", {
                            {"type", "string"},
                            {"enum", {"add", "subtract", "multiply", "divide"}},
                            {"description", "The operation to perform"}
                        }}
                    }},
                    {"required", {"a", "b", "operation"}}
                }}
            }}
        },
        {
            {"type", "function"},
            {"function", {
                {"name", "get_weather"},
                {"description", "Get current weather for a location"},
                {"parameters", {
                    {"type", "object"},
                    {"properties", {
                        {"location", {{"type", "string"}, {"description", "City name"}}}
                    }},
                    {"required", {"location"}}
                }}
            }}
        }
    });

    /* Step 4: User's request */
    std::string user_query = "What time is it, and what's 42 multiplied by 17?";
    std::cout << "User: " << user_query << "\n\n";

    /* Step 5: Build initial request */
    CURL *curl = curl_easy_init();
    if (curl == nullptr) {
        curl_global_cleanup();
        return 1;
    }

    json request_body = {
        {"model", "meta-llama/llama-4-scout-17b-16e-instruct"},
        {"messages", {
            {{"role", "user"}, {"content", user_query}}
        }},
        {"tools", tools},
        {"tool_choice", "auto"}
    };

    std::string json_payload = request_body.dump();

    /* Headers */
    struct curl_slist *headers = nullptr;
    headers = curl_slist_append(headers, "Content-Type: application/json");
    std::string auth_header = "Authorization: Bearer " + api_key;
    headers = curl_slist_append(headers, auth_header.c_str());

    /* Configure and send */
    std::string response_data;
    curl_easy_setopt(curl, CURLOPT_URL, "https://api.demeterics.com/groq/v1/chat/completions");
    curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);
    curl_easy_setopt(curl, CURLOPT_POSTFIELDS, json_payload.c_str());
    curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
    curl_easy_setopt(curl, CURLOPT_WRITEDATA, &response_data);

    std::cout << "Sending request to AI...\n";
    CURLcode res = curl_easy_perform(curl);

    if (res == CURLE_OK) {
        try {
            json response = json::parse(response_data);

            if (response.contains("choices") && !response["choices"].empty()) {
                auto message = response["choices"][0]["message"];

                if (message.contains("tool_calls") && message["tool_calls"].is_array()) {
                    std::cout << "\nAI wants to call " << message["tool_calls"].size()
                              << " tool(s):\n\n";

                    /* Process each tool call */
                    for (const auto &call : message["tool_calls"]) {
                        std::string name = call["function"]["name"];
                        std::string args_str = call["function"]["arguments"];
                        json args = json::parse(args_str);

                        std::cout << "Tool: " << name << "\n";
                        std::cout << "Args: " << args_str << "\n";

                        /* Execute the tool */
                        std::string result;

                        if (name == "get_current_time") {
                            result = get_current_time();
                        }
                        else if (name == "calculate") {
                            double a = args["a"];
                            double b = args["b"];
                            std::string op = args["operation"];
                            result = calculate(a, b, op);
                        }
                        else if (name == "get_weather") {
                            std::string location = args["location"];
                            result = get_weather(location);
                        }

                        std::cout << "Result: " << result << "\n\n";
                    }
                } else {
                    /* No tool calls - direct response */
                    if (message.contains("content") && !message["content"].is_null()) {
                        std::cout << "\nAI Response: " << message["content"] << "\n";
                    }
                }
            }
        } catch (const json::exception &e) {
            std::cerr << "Parse error: " << e.what() << "\n";
        }
    } else {
        std::cerr << "Error: " << curl_easy_strerror(res) << "\n";
    }

    /* Clean up */
    curl_slist_free_all(headers);
    curl_easy_cleanup(curl);
    curl_global_cleanup();

    /* Educational info */
    std::cout << "=== TOOL USE PATTERN ===\n\n";
    std::cout << "1. Define tools with JSON schema\n";
    std::cout << "2. Send user query + tool definitions\n";
    std::cout << "3. AI returns tool_calls if needed\n";
    std::cout << "4. Execute tools locally\n";
    std::cout << "5. Send results back to AI\n";
    std::cout << "6. AI generates final response\n\n";

    std::cout << "USE CASES:\n";
    std::cout << "  - Web search agents\n";
    std::cout << "  - Database query agents\n";
    std::cout << "  - API integration\n";
    std::cout << "  - Code execution\n";
    std::cout << "  - Multi-step reasoning\n";

    return 0;
}

/*
 * C++ CONCEPTS:
 *
 * std::to_string(number)
 *   Converts number to string.
 *
 * json::array({...})
 *   Creates a JSON array with initializer list.
 *
 * json::object()
 *   Creates an empty JSON object.
 *
 * TOOL CALLING NOTES:
 *
 * TOOL DEFINITION SCHEMA:
 *   - name: Function identifier
 *   - description: What the function does
 *   - parameters: JSON Schema for arguments
 *
 * AI DECIDES:
 *   - Which tool to call (or none)
 *   - What arguments to pass
 *   - Can call multiple tools
 */
